### PR TITLE
DOC: Refactor numeric index docs

### PR DESCRIPTION
**Pull request by gfyoung**
Refactor `NumericIndex` docs to avoid duplicate documentation.  xref #14937.

### CODE REVIEW

**Comment by jreback** [comment]
nit: needs a blank like before ``Paramaters``. need to wrap the line for Int64Index.

```

In [1]: pd.Int64Index?
Init signature: pd.Int64Index(data=None, dtype=None, copy=False, name=None, fastpath=False)
Docstring:     
Immutable ndarray implementing an ordered, sliceable set. The basic object
storing axis labels for all pandas objects. Int64Index is a special case
of `Index` with purely integer labels. This is the default index type used by the DataFrame and Series ctors when no explicit index is provided by the user.
Parameters
----------
data : array-like (1-dimensional)
dtype : NumPy dtype (default: int64)
copy : bool
    Make a copy of input ndarray
name : object
    Name to be stored in the index
Notes
-----
An Index instance can **only** contain hashable objects.
File:           ~/pandas/pandas/indexes/numeric.py
Type:           type

In [2]: pd.Float64Index?
Init signature: pd.Float64Index(data=None, dtype=None, copy=False, name=None, fastpath=False)
Docstring:     
Immutable ndarray implementing an ordered, sliceable set. The basic object
storing axis labels for all pandas objects. Float64Index is a special case
of `Index` with purely integer labels. 
Parameters
----------
data : array-like (1-dimensional)
dtype : NumPy dtype (default: float64)
copy : bool
    Make a copy of input ndarray
name : object
    Name to be stored in the index
Notes
-----
An Index instance can **only** contain hashable objects.
File:           ~/pandas/pandas/indexes/numeric.py
Type:           type
```

**Comment by gfyoung** [comment]
@jreback : Made the requested doc changes.  Ready to merge if there are no other concerns.

File: pandas/indexes/base.py [hunk 1]
```diff
@@ -1128,17 +1128,16 @@ def is_mixed(self):
...
         return self.inferred_type in ['integer', 'mixed-integer']
 
-    # validate / convert indexers
-    def _convert_scalar_indexer(self, key, kind=None):
-        """
-        convert a scalar indexer
-
+    _index_shared_docs['_convert_scalar_indexer'] = """
+        Convert a scalar indexer.
         Parameters
```
Comment (jorisvandenbossche) [inline pandas/indexes/base.py]: also a blank line missing above 'Parameters' here I think

File: pandas/indexes/numeric.py [hunk 2]
```diff
@@ -90,27 +74,36 @@ def _assert_safe_casting(cls, data, subarr):
...
-class Int64Index(NumericIndex):
-    """
+_num_index_shared_docs['class_descr'] = """
     Immutable ndarray implementing an ordered, sliceable set. The basic object
-    storing axis labels for all pandas objects. Int64Index is a special case
-    of `Index` with purely integer labels. This is the default index type used
-    by the DataFrame and Series ctors when no explicit index is provided by the
-    user.
+    storing axis labels for all pandas objects. %(klass)s is a special case
+    of `Index` with purely integer labels. %(extra)s
```
Comment (jorisvandenbossche) [inline pandas/indexes/numeric.py]: "purely integer labels" -> will also be used for Float64Index?

**Review by jorisvandenbossche — COMMENTED** [review]
+1 on sharing docstrings for public functions (eg the classes in this PR), but I am not sure we should also do this for private helper functions. 
As the docstrings there are in the first place helpful for developers when looking and trying to understand the code. That the docstring is right there next to the function when looking at it in your IDE can be very helpful.

**Comment by gfyoung** [comment]
@jorisvandenbossche : I think sharing docstrings across private functions is fine.  I get more confused by the inconsistencies across written-out docstrings for the same functions, whether private or public.  Not to mention, there are already numerous private functions in `base.py` that docstring-shared.

File: pandas/indexes/base.py (hunk 1 above)
Comment (gfyoung) [inline pandas/indexes/base.py]: Good catch.  Done.

File: pandas/indexes/numeric.py (hunk 2 above)
Comment (gfyoung) [inline pandas/indexes/numeric.py]: Good catch.  Done.

**Comment by gfyoung** [comment]
@jreback , @jorisvandenbossche : If there are no other concerns, I think this is good to merge.

**Comment by gfyoung** [comment]
@jreback , @jorisvandenbossche : If there are no other concerns, I think this is good to merge.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**